neuromation/api/object_storage.py

```diff
@@ -0,0 +1,401 @@
+            return [_obj_status_from_bucket(bucket) for bucket in res]
+
+    async def list_objects(
+        self, bucket_name: str, prefix: str = "", recursive: bool = False,
+    ) -> List[ObjStatus]:
+        url = self._config.obj_url / "o" / bucket_name
+        auth = await self._config._api_auth()
+
+        query = {"recursive": str(recursive).lower()}
+        if prefix:
+            query["prefix"] = prefix
+        url = url.with_query(query)
+
+        contents: List[ObjStatus] = []
+        common_prefixes: List[ObjStatus] = []
+        while True:
+            async with self._core.request("GET", url, auth=auth) as resp:
+                res = await resp.json()
+                contents.extend(
+                    [_obj_status_from_key(bucket_name, key) for key in res["contents"]]
+                )
+                common_prefixes.extend(
+                    [
+                        _obj_status_from_prefix(bucket_name, prefix)
+                        for prefix in res["common_prefixes"]
+                    ]
+                )
+                if res["is_truncated"] and res["contents"]:
+                    start_after = res["contents"][-1]["key"]
+                    url = url.with_query(start_after=start_after)
+                else:
+                    break
+        return common_prefixes + contents
+
+    async def glob_objects(self, bucket_name: str, pattern: str) -> List[ObjStatus]:
+        pattern = pattern.lstrip("/")
+        parts = pattern.split("/")
+        # Limit the search to prefix of keys
+        prefix = ""
+        for part in parts:
+            if _has_magic(part):
+                break
+            else:
+                prefix += part + "/"
+
+        match = re.compile(fnmatch.translate(pattern)).fullmatch
+        res = []
+        for obj_status in await self.list_objects(bucket_name, prefix=prefix):
+            if match(obj_status.name):
+                res.append(obj_status)
+        return res
+
+    async def head_object(self, bucket_name: str, key: str) -> ObjStatus:
+        url = self._config.obj_url / "o" / bucket_name / key
+        auth = await self._config._api_auth()
+
+        async with self._core.request("HEAD", url, auth=auth) as resp:
+            type_ = (
+                FileStatusType.DIRECTORY if key.endswith("/") else FileStatusType.FILE
+            )
+            modification_time = 0
+            if "Last-Modified" in resp.headers:
+                timetuple = parsedate(resp.headers["Last-Modified"])
+                if timetuple is not None:
+                    modification_time = int(time.mktime(timetuple))
+            return ObjStatus(
+                path=key,
+                type=type_,
+                size=resp.content_length or 0,
+                modification_time=modification_time,
+                bucket_name=bucket_name,
+            )
+
+    async def fetch_object(self, bucket_name: str, key: str) -> AsyncIterator[bytes]:
```

Comment:
GetObject in botocore returns the stream as part of the response. I was hesitant to return an object that would take responsibility for yet not closed request `Object(resp=resp)`. But the response headers and metadata are not accessible using this API, which bothers me.

Reply:
Does `Tuple[ObjectStatus, AsyncIterator[bytes]]` solve your needs?
Or maybe a dataclass instead of tuple?